### PR TITLE
Massive performance improvements + code cleanup

### PR DESCRIPTION
**Pull request by ofekih**
Hey, saw your video and that got me interested in this problem. I also heard you saying that you know that your code wasn't the cleanest, so I spent a few minutes cleaning it up.

Note: This uses python3 stuff, so be sure to run this with python3 if you want to test.

1)

Basically the biggest bottleneck in your code was the way you counted how many times a number showed up.

Counting numbers in a list is O(n), which is pretty bad, and it's done O(n) times, so that was O(n²) time spent on it. Using a dictionary, with an amortized access time of O(1), this can be reduced to O(1) per iteration, with O(n) iterations, so O(n) time spent on it. In short, the bottleneck for the algorithm can be asymptotically sped up, resulting in much better performance (you can now go up to 501 in a small fraction of a second).

I also think using a dictionary is more intuitive, since you only need to keep track of one collection (the dictionary) as opposed to 5 objects (`list_numbers`, `set_numbers`, `how_many_times`, `which_ones`, and finally, the dictionary `final_info`).

2)

Next, the factorial part of your code was not the bottleneck, but it was still slow. Huge factorials were being calculated and divided many times for the nCr calculation. It should be faster to simply use the library implementation (`math.comb(n, r)`), which I think avoids doing the whole calculation.

However, since we are doing many calculations of nCr, it would be even faster to use the previous value of nCr to calculate the next value. Feel free to verify by sketching out a few terms, but you can show that `comb(n, r) = comb(n, r - 1) * (n - r + 1) / r`. Basically, you can just see how the factorials change from one term to the next and just do a single multiplication and division instead of recalculating all the factorials.

Again, this shouldn't be the bottleneck, and I'd be happy to change it back to something like `math.comb(n, r)` if you prefer the simplicity, but I think it's a pretty simple and elegant solution to include.

3)

I cleaned up the code a bit. I moved the main logic to an `if __name__ == '__main__'` block per python convention, and I made search just return a list of tuples of the matches and frequencies it found rather than print out anything. The printing out is done in the if name main block. I also added a `minimum_counts` argument in case you want to search for values that show up at least 8 times or something.

4)

I added some simple elapsed time calculation.

Cheers

### CODE REVIEW

File: search.py [hunk 1]
```diff
@@ -1,36 +1,49 @@
...
+
+        matches.append((entry, frequency))
+
+    return matches
+
+
+if __name__ == '__main__':
+    start_time = time.time()
+
+    matches = search(3001, 5)
```
Comment (ofekih) [inline search.py]: runs in ~5 seconds on my computer, I can change this to anything. also the '5' is optional, you can just say: `search(3001)`

Comment (achampion) [reply search.py]: You can use parts of the standard library to reduce the work, namely `collections` and `itertools`, e.g.:

    import itertools as it
    from collections import Counter

    def search(limit):
        c = Counter(it.chain.from_iterable(it.accumulate(range(3, n - 1),
                                                         lambda x, r: x * (n - r + 1) // r,
                                                         initial=n * (n - 1) // 2) for n in range(4, limit)))
        return it.takewhile(lambda i: i[1] >= 4, c.most_common())

    for entry, frequency in search(3000):
        print(f"The number {entry} appears at least {frequency+2} times across the whole triangle")

**Comment by achampion** [comment]
You can use parts of the standard library to reduce the work, namely `collections` and `itertools`, e.g.:

    import itertools as it
    from collections import Counter

    def search(limit):
        c = Counter(it.chain.from_iterable(it.accumulate(range(3, n - 1),
                                                         lambda x, r: x * (n - r + 1) // r,
                                                         initial=n * (n - 1) // 2) for n in range(4, limit)))
        return it.takewhile(lambda i: i[1] >= 4, c.most_common())

    for entry, frequency in search(3000):
        print(f"The number {entry} appears at least {frequency+2} times across the whole triangle")

File: search.py (hunk 1 above)
Comment (achampion) [reply search.py]: The `math.factorial()` code was the biggest consumer of resources after `list.count()`

Comment (ofekih) [reply search.py]: Hey @achmpion, thanks for your review. As I mentioned in my (admittedly long) description above, the factorial code is not the bottleneck, the O(n²) brought by `list.count` is. Performance gains by improving the factorial code (by perhaps using `math.comb` or my above smart combinatorics code) are negligible before removing `list.count`.

So I think your comment is a bit misleading.

I would rephrase: "The `math.factorial()` code was the biggest consumer of resources after `list.count()`" to something less misleading, like: "`list.count()` is by far the bottleneck and biggest consumer of resources, after which `math.factorial()` starts to become non-negligible."

**Comment by ofekih** [comment]
> You can use parts of the standard library to reduce the work, namely `collections` and `itertools`, e.g.:
> 
> ```
> import itertools as it
> from collections import Counter
> 
> def search(limit):
>     c = Counter(it.chain.from_iterable(it.accumulate(range(3, n - 1),
>                                                      lambda x, r: x * (n - r + 1) // r,
>                                                      initial=n * (n - 1) // 2) for n in range(4, limit)))
>     return it.takewhile(lambda i: i[1] >= 4, c.most_common())
> 
> for entry, frequency in search(3000):
>     print(f"The number {entry} appears at least {frequency+2} times across the whole triangle")
> ```

Sweet idea. There are many ways to reduce work, I was considering using reduce and some other functools. Your above suggestion is particularly clever. It was important for me to keep the code simple and understandable for demonstration purposes, since I think that was the primary purpose of this repository.

While your code is more clever, it is also less clear what it does and how it works. I like to think that my pull request actually simplifies the code and makes it easier to understand, while providing an asymptotically significant improvement.

Eyeballing your code, it looks like if anything it will perform worse (asymptotically) than my current code, and looks less clear in my opinion. `most_common` likely performs a sort, and that sort is likely O(n log n). My solution should just be O(n), where `n` is the number of integers checked.

Per my expectations, your solution seems to run faster when the limit is ~2000, but mines runs faster when the limit is ~3000+. My code is around twice as fast after 4500, and the difference only grows.

I commend your clever engineering, but I prefer clearer to understand and asymptotically more performant code than albeit clever but nevertheless obscure code that saves a few lines.

**Comment by achampion** [comment]
Understood and there is always a balance to be struck between how much to leverage libraries (especially the standard library) and how much to write yourself.

`itertools.accumulate()` is great for anytime you are repeatedly applying an accumulating function.  
`collections.Counter()` does exactly what it says on the tin and counts the number of occurrences of things (multiset)

I do agree the `itertools.chain.from_iterable()` looks complicated but this could all be replace with:

    c = Counter()
    for n in range(4, limit):
        c.update(it.accumulate(range(3, n - 1), lambda x, r: x * (n - r + 1) // r, initial=n * (n - 1) // 2))

And if you wanted you could pull out the `lambda`:

    c = Counter()
    for n in range(4, limit):
        nCr_incr = lambda x, r: x * (n - r + 1) // r
        c.update(it.accumulate(range(3, n - 1), nCr_incr, initial=nCr_incr(n, 2)))

Yes `itertools.takewhile()` is `O(n)`. `Counter.most_common()` is `O(n log k)` (heapsort), good catch I was originally just looking at the top results. But you can treat the `Counter` as a plain `dict` and just filter all values:

    return filter(lambda i: i[1] >= 4, c.items())   # O(n)

`most_common()` would be a better solution if you wanted to limit to the `k` most common, where k is small.   
This brings the timing to about the same (a slight improvement for using the standard library which is written in C):

    my_search(4500): CPU times: user 14.8 s, sys: 745 ms, total: 15.5 s
    your_search(4500): CPU times: user 16.3 s, sys: 774 ms, total: 17.1 s

**Comment by ofekih** [comment]
> Understood and there is always a balance to be struck between how much to leverage libraries (especially the standard library) and how much to write yourself.
> 
> `itertools.accumulate()` is great for anytime you are repeatedly applying an accumulating function.
> `collections.Counter()` does exactly what it says on the tin and counts the number of occurrences of things (multiset)
> 
> I do agree the `itertools.chain.from_iterable()` looks complicated but this could all be replace with:
> 
> ```
> c = Counter()
> for n in range(4, limit):
>     c.update(it.accumulate(range(3, n - 1), lambda x, r: x * (n - r + 1) // r, initial=n * (n - 1) // 2))
> ```
> 
> And if you wanted you could pull out the `lambda`:
> 
> ```
> c = Counter()
> for n in range(4, limit):
>     nCr_incr = lambda x, r: x * (n - r + 1) // r
>     c.update(it.accumulate(range(3, n - 1), nCr_incr, initial=nCr_incr(n, 2)))
> ```

That looks much better, but I don't really see the advantage of this. Combining the `nCr` update to one line is probably better though, I should change that at some point. The way I look at it is: lines of code don't cost money. These two solutions (for populating the Counter/dict) should run in a nearly identical amount of time. Everyone understands simple loops. There is little advantage to using library code here.

> Yes `itertools.takewhile()` is `O(n)`. `Counter.most_common()` is `O(n log k)` (heapsort), good catch I was originally just looking at the top results. But you can treat the `Counter` as a plain `dict` and just filter all values:
> 
> ```
> return filter(lambda i: i[1] >= 4, c.items())   # O(n)
> ```

Oh that's interesting, didn't know much about `most_common`. Makes sense that it's `(O(n log k))`. And you're right, `filter` is better, since even heapsort needs to do 1 comparison for each number (technically `n-1` comparisons) before spitting out the largest, and all this filter is doing is literally a single integer comparison.

> `most_common()` would be a better solution if you wanted to limit to the `k` most common, where k is small.


> This brings the timing to about the same (a slight improvement for using the standard library which is written in C):
> 
> ```
> my_search(4500): CPU times: user 14.8 s, sys: 745 ms, total: 15.5 s
> your_search(4500): CPU times: user 16.3 s, sys: 774 ms, total: 17.1 s
> ```

Yeah, understandable that yours runs faster now. `filter` indeed resolves the performance issue. Again, personally, I think the audience (if this is merged) might be largely non-technical, and would just wanna see simple code with large comments. The audience might also not be comfortable with `python`, so I think limiting standard library usage when it isn't clear is probably better. `filter` is probably ok, since many languages have it, but `Counter` might be a bit more python-specific.

Sweet that you got your algorithm to outperform mine :P  With that `filter` change it's probably asymptotically the same is mine, meaning it will outperform mine forever ;(  I'm still partial to a less advanced approach, since I don't think performance is the most important (as long as it is not asymptotically worse).

I have work, but afterwards today if I'm bored out of my mind because of covid, I could whip something up in c++ that I suspect would bring performance improvements. A few ideas (again, I think for the purposes of this repo simpler is better, but it's always fun to over-optimize):

1) Parallelization: This problem is highly parallelizable, and there are a few ways to do it (either update at the end or a few times in the middle)

2) Custom dictionary/map: Probably a performance improvement for larger `n`'s if we can delete old values, perhaps when there is a collision. Once our loop reaches a row, we know that any number up 'til that row `r` choose `2` will never be seen again. So if we get a collision we can probably just delete the colliding value. Or we can mess with other ideas, but custom dict means more control.

Probably more ideas, but I have to go

**Comment by rioj7** [comment]
Using achampion's refactor I learned new collections and itertools functions and I like the new found number

3537835171522765057006983148520718494957187357011427136691375227388082606684583032666088334962061461901090477513197821330000906170565587040820236444389470701575515092325417606033095416151914090271577807800

a lot

**Comment by achampion** [comment]
Nice, if you are interested here's the gist for how I would implement it, it's slightly slower but the same `O()`: https://gist.github.com/achampion/cd0ac77c5b9a3716936445623bccd8d3
